gpt4all-backend/gptj.cpp

```diff
@@ -890,159 +890,50 @@ size_t GPTJ::restoreState(const uint8_t *src)
     return gptj_set_state_data(d_ptr->model, &d_ptr->rng, src);
 }
 
-void GPTJ::prompt(const std::string &prompt,
-        std::function<bool(int32_t)> promptCallback,
-        std::function<bool(int32_t, const std::string&)> responseCallback,
-        std::function<bool(bool)> recalculateCallback,
-        PromptContext &promptCtx) {
-
-    if (!isModelLoaded()) {
-        std::cerr << "GPT-J ERROR: prompt won't work with an unloaded model!\n";
-        return;
-    }
-
-    // tokenize the prompt
-    std::vector<gpt_vocab::id> embd_inp = ::gpt_tokenize(d_ptr->vocab, prompt);
-
-    // save the context size
-    promptCtx.n_ctx = d_ptr->model->hparams.n_ctx;
+std::vector<LLModel::Token> GPTJ::tokenize(const std::string &str)
+{
+    return ::gpt_tokenize(d_ptr->vocab, str);
+}
 
-    if ((int) embd_inp.size() > promptCtx.n_ctx - 4) {
-        responseCallback(-1, "ERROR: The prompt size exceeds the context window size and cannot be processed.");
-        std::cerr << "GPT-J ERROR: The prompt is" << embd_inp.size() <<
-            "tokens and the context window is" << promptCtx.n_ctx << "!\n";
-        return;
-    }
+LLModel::Token GPTJ::sampleToken(PromptContext &promptCtx)
+{
+    const size_t n_prev_toks = std::min((size_t) promptCtx.repeat_last_n, promptCtx.tokens.size());
+    return gpt_sample_top_k_top_p(d_ptr->model->hparams.n_vocab,
+        promptCtx.tokens.data() + promptCtx.tokens.size() - n_prev_toks,
+        n_prev_toks,
+        promptCtx.logits,
+        promptCtx.top_k, promptCtx.top_p, promptCtx.temp,
+        promptCtx.repeat_penalty,
+        d_ptr->rng);
+}
 
-    promptCtx.n_predict = std::min(promptCtx.n_predict, promptCtx.n_ctx - (int) embd_inp.size());
-    promptCtx.n_past = std::min(promptCtx.n_past, promptCtx.n_ctx);
+std::string_view GPTJ::tokenToString(Token id)
+{
+    return d_ptr->vocab.id_to_token[id];
+}
 
+bool GPTJ::evalTokens(PromptContext &ctx, const std::vector<int32_t> &tokens)
+{
     // determine the required inference memory per token:
     static bool initialized = false;
-    static std::vector<gpt_vocab::id> p_instruct;
-    static std::vector<gpt_vocab::id> r_instruct;
     if (!initialized) {
```

Comment:
Is there a chance that this function could be run from parallel threads?

Reply:
The client doesn't do that. The bindings could try and do it.

Reply:
Either way, it is a good point we should look at thread safety if someone were trying to load multiple models in RAM and run them from inference at same time, but it isn't germane to this PR.

Reply:
Right, now that I look at it those racy `initialized` flags were already there before the PR.